content/post/quality-and-speed.md

```diff
@@ -0,0 +1,134 @@
+date: 2020-08-31T09:00:03+01:00
+categories: ["Ramblings"]
+small_images:
+- "/img/development.png"
+---
+
+<img src='/img/development.png' style='float:right; width:200px;margin-left:15px'/>
+
+From the article [Is High Quality Software Worth the
+Cost?](https://martinfowler.com/articles/is-quality-worth-cost.html) I found a
+particular quote super interesting:
+
+> Countless times I've talked to development teams who say "they (management)
+> won't let us write good quality code because it takes too long". Developers
+> often justify attention to quality by justifying through the need for proper
+> professionalism. But this moralistic argument implies that this quality comes
+> at a cost - dooming their argument.
+
+This is something I can relate with. For too long I believed that I needed more
+time to get things done better. And that's true to a point. If we're doing
+craftsmanship, the more time we put into something, the more polished it will be.
+Right?
+
+<div style='clear:both'></div>
+<!--more-->
+
+## Is time the main requirement?
+
+Let me post here something that was made for my youngest daughter's birthday.
+
+<img src='/img/snowhite.jpg' />
+
+One was [done by my
+sister](https://www.instagram.com/p/CEfSjD_jkFS/?utm_source=ig_web_copy_link)
+and the other by my niece. It's easy to see that we have a piece of work that is more
+polished than the other one. The structure is similar but there are several
+quality differences that are easy to spot.
+
+Now, my sister did the best version of Snow White _much faster_ than my niece.
+We can argue that the comparison is not fair, after all my sister has _more
+experience and skills_ than my niece.
+
+But I can also argue that _time_ is not the main requirement for quality. I
+could say that my niece would not be able to achieve better quality even if she
+had one more week or month.
+
+## Is skill set the main requirement?
+
+Okay, now lets play the devil's advocate on the skill issue. If we look at the
+figure, we could say for example that the face is more detailed than the skirt.
+So it's not just _skill_. With more time my sister could have done a better job
+on the whole figure.
+
+On software engineering we deal with a lot of different modules. We may have
+ones that are more critical and need more attention. And we have other
+interesting challenges. We could actually need to build several figures, we
+could have several people working and need to coordinate them and make sure they
+are all aligned and using the same standards, etc.
+
+All this with a time constraint that wouldn't allow for a proper figure to be
+built.
+
+And lets not forget change requests. We can all relate with building a snow
+white's face that is top notch and then the customer asks for Little Red Hidding
+Hood.
+
+## Where's the silver bullet?
+
+Time does impact quality. If we gave a deadline of 2 minutes, maybe both my
+sister and my niece would present us with the same output. On the other hand I
```

Comment:
I would disagree with this. Your sister has far more experience, has another kind of maturity, and, as an adult, handles pressure a different way.

We can argue that the final result would be bad in both cases, but I would assume that your sister would still have a better result.

This reminds me of a speed drawing challenge (https://vimeo.com/226508728). Independently of your level of skill, if you don't have time, you will do a lousy job.

In software development, I see several things that happen that can be related to skill, other not so much:
- Deadlines are given by others - In this case, engineers might do a good job, but if you have 1 day to do something that would take a week, there are no miracles. Quality will suffer and problems will arise (or not, you can be lucky :) ).
- Deadlines are given by engineers - In this case, we can put the skill into the equation, because experienced developers will take into consideration that the task is not only writing the code. Is also guaranteeing the quality of that code. And this is something that some organizations don't understand that delivering something quickly is not necessarily deliver something faster. We need to spend some time at the beginning (setting up all the quality pipelines and testing strategies) and during (ensuring the quality of our code) to be faster at the end (meaning delivering a high-quality product that our clients want to use and we don't spend months/years handling the tech debt/problems that were created during the development phase. Of course, there might be compromises but, in my opinion, we should compromise on functionality and not on quality.